### PR TITLE
Scripting on .NET Core 1.1

### PR DESCRIPTION
**Pull request by tmat**
The resolution of assembly names specified by the scripting host or by the script via ```#r``` directive uses GAC on Desktop FX, but wasn't implemented on CoreCLR. This change implements the resolution by looking into a TPA (Trusted Platform Assembly) list provided by CoreCLR. It also adjust the ```InteractiveAssemblyLoader``` to the change in CoreCLR assembly loader between versions 1.0 and 1.1 that broke it. Finally, it enables all portable scripting tests on CoreCLR; they were not running previously due to the missing Framework assembly name resolution. 

**Customer scenario**

```#r``` didn't resolve Framework assembly names when running on CoreCLR.
Scripting API didn't work on CoreCLR 1.1 due to a breaking change in the CLR assembly loader.

**Bugs this fixes:** 

https://github.com/dotnet/roslyn/issues/15352
https://github.com/dotnet/roslyn/issues/5308
https://github.com/dotnet/roslyn/issues/5167

**Workarounds, if any**

None.

**Risk**

Small. 

**Performance impact**

Low.

**Is this a regression from a previous update?**

Yes, in CoreCLR. Not in Roslyn.

**Root cause analysis:**

We were not running any scripting tests on .NET Core. We also haven't updated to .NET Core 1.1 until now. This change enables scripting tests on .NET Core.

**How was the bug found?**

Customer reported.

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler @dotnet/roslyn-interactive Please review.

**Comment by AlekseyTs** [comment]
Changes under compilers LGTM

**Comment by tmat** [comment]
@nguerrera FYI.

File: src/Interactive/CsiCore/Csi.cs [hunk 1]
```diff
@@ -17,14 +15,18 @@ internal static int Main(string[] args)
         {
             try
             {
-                var responseFile = Path.Combine(AppContext.BaseDirectory, InteractiveResponseFileName);
+                // Note that AppContext.BaseDirectory isn't necessarily the directory containing csi.exe.
+                // For example, when executed via corerun it's the directory containing corerun.
+                string csiDirectory = Path.GetDirectoryName(typeof(Csi).GetTypeInfo().Assembly.ManifestModule.FullyQualifiedName);
```
Comment (nguerrera) [inline src/Interactive/CsiCore/Csi.cs]: I'm not opposed to the change, but that seems like a bug in corerun. Is there a better example to use in the comment than an internal coreclr host?

File: src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs [hunk 2]
```diff
@@ -567,5 +583,541 @@ public void HostObjectInInMemoryAssembly()
...
+
+            bool exceptionSeen = false;
+            try
+            {
+                await s2.ContinueWithAsync($@"var l2 = new Lib2();");
+            }
+            catch (FileLoadException fileLoadEx) when (fileLoadEx.InnerException is InteractiveAssemblyLoaderException)
+            {
+                exceptionSeen = true;
+            }
```
Comment (nguerrera) [inline src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs]: Just curious, is there not a way to do Assert.Throws on async method? Should we log xunit issue?

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs [hunk 3]
```diff
@@ -26,10 +27,55 @@ public class CommandLineRunnerTests : TestBase
...
 
+        private static IEnumerable<string> GetReferences()
+        {
+            if (GacFileResolver.IsAvailable)
+            {
+                // keep in sync with list in csi.rsp
+                yield return "System";
+                yield return "System.Core";
+                yield return "Microsoft.CSharp";
+                yield return "System.ValueTuple.dll";
```
Comment (nguerrera) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: Nit: personal preference, but why the column of yield return over new[] {...}?

File: src/Scripting/CSharpTest/ObjectFormatterTests.cs [hunk 4]
```diff
@@ -418,7 +418,16 @@ public void DebuggerProxy_FrameworkTypes_IEnumerable()
 
             obj = Enumerable.Range(0, 10);
             str = s_formatter.FormatObject(obj, SingleLineOptions);
-            Assert.Equal("RangeIterator { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }", str);
+
+            // the implementation differs between .NET Core and .NET FX
```
Comment (nguerrera) [inline src/Scripting/CSharpTest/ObjectFormatterTests.cs]: Yuck. Bug?

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs [hunk 5]
```diff
@@ -106,28 +131,90 @@ public override ImmutableArray<PortableExecutableReference> ResolveReference(str
...
+                            string fileName = Path.GetFileNameWithoutExtension(path);
+                            if (fileName.EndsWith(".ni", StringComparison.OrdinalIgnoreCase))
+                            {
+                                fileName = fileName.Substring(0, fileName.Length - ".ni".Length);
+                            }
+
+                            // last one wins:
+                            set[fileName] = path;
+                        }
+                    }
```
Comment (nguerrera) [inline src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: The scope of this catch is big enough to trigger my concern that a bug could be eaten. Can you reduce the scope of tighten the exception types caught?

File: src/Scripting/CoreTestUtilities/RedirectedOutput.cs [hunk 6]
```diff
@@ -2,16 +2,21 @@
...
     public sealed class OutputRedirect : IDisposable
     {
+        private static readonly object s_guard = new object();
+
         private readonly TextWriter _oldOut;
         private readonly StringWriter _newOut;
 
         public OutputRedirect(IFormatProvider formatProvider)
         {
+            Monitor.Enter(s_guard);
```
Comment (nguerrera) [inline src/Scripting/CoreTestUtilities/RedirectedOutput.cs]: Why not lock?

Comment (tannergooding) [inline src/Scripting/CoreTestUtilities/RedirectedOutput.cs]: I'm guessing the Monitor is used because the release happens when the object is disposed. I also feel like there are better ways to do this than using a monitor to lock over the lifetime of the object.

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs (hunk 3 above)
Comment (tannergooding) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: I would guess because this is non-allocating.

File: src/Interactive/CsiCore/Csi.cs (hunk 1 above)
Comment (tmat) [inline src/Interactive/CsiCore/Csi.cs]: I'm not sure. The doc on ```AppContext.BaseDirectory``` says 

> Gets the base directory that the assembly resolver uses to probe for assemblies.

That doesn't seem to guarantee that it's the directory ```csi.exe``` is located in.

File: src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs (hunk 2 above)
Comment (tmat) [inline src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs]: Don't remember. This test was already there, it's just moved. I think there is ```Assert.Throws<T>``` with async method but  there isn't one with conditional filter (note that here we check for the inner exception.

File: src/Scripting/CSharpTest/ObjectFormatterTests.cs (hunk 4 above)
Comment (tmat) [inline src/Scripting/CSharpTest/ObjectFormatterTests.cs]: Not a bug, just a different implementation. The object display is similar to a debugger display in immediate window, just implemented thru Reflection. It sometimes captures an implementation detail of the code. In one of the cases the method is implemented thru iterator (the object display actually decodes the compiler generated name of the iterator method), the other is implemented thru struct, I think.

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs (hunk 5 above)
Comment (tmat) [inline src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: Fair enough. I guess I can use non-throwing ```FileNameUtilities```

File: src/Scripting/CoreTestUtilities/RedirectedOutput.cs (hunk 6 above)
Comment (tmat) [inline src/Scripting/CoreTestUtilities/RedirectedOutput.cs]: @tannergooding Yes, the lock is held until it's disposed. What better ways? (Note that this is test code.)

File: src/Interactive/CsiCore/Csi.cs (hunk 1 above)
Comment (jkotas) [inline src/Interactive/CsiCore/Csi.cs]: corerun does not set it to anything, and so it has the default value. Yes, it may be nice to fix it in corerun.

File: src/Scripting/CoreTestUtilities/RedirectedOutput.cs (hunk 6 above)
Comment (tannergooding) [inline src/Scripting/CoreTestUtilities/RedirectedOutput.cs]: Using a TestCollection to guarantee the tests in question (and only those tests) are serialized wrt everything else.

File: src/Compilers/Shared/CoreClrShim.cs [hunk 7]
```diff
@@ -15,5 +16,15 @@ internal static class AssemblyLoadContext
             internal static readonly Type Type = ReflectionUtilities.TryGetType(
                "System.Runtime.Loader.AssemblyLoadContext, System.Runtime.Loader, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
         }
+
+        internal static class AppContext
+        {
+            internal static readonly Type Type = ReflectionUtilities.TryGetType(
+                "System.AppContext, System.AppContext, Version=4.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a");
```
Comment (jaredpar) [inline src/Compilers/Shared/CoreClrShim.cs]: Does this version string need to line up with the version in project.json?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs [hunk 8]
```diff
@@ -262,11 +262,13 @@ private void ProcessStarting(bool initialize)
 
         private static MetadataReferenceResolver CreateMetadataReferenceResolver(IMetadataService metadataService, ImmutableArray<string> searchPaths, string baseDirectory)
         {
+            // TODO: We need to do reference resolution in the remote process (InteractiveHost.exe), in order to support other platforms than Desktop FX
```
Comment (jaredpar) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Bug number?

File: src/Interactive/CsiCore/Csi.cs (hunk 1 above)
Comment (jaredpar) [inline src/Interactive/CsiCore/Csi.cs]: Under what circumstances was this change necessary?  We've been under the same assumption about `AppContext.BaseDirectory` in the compiler code.

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs [hunk 9]
```diff
@@ -330,15 +376,15 @@ public void Args_Script1()
 ", runner.Console.Out.ToString());
         }
 
-        [Fact]
+        [ConditionalFact(typeof(WindowsOnly)), WorkItem(15860, "https://github.com/dotnet/roslyn/issues/15860")]
```
Comment (jaredpar) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: What changed that necessitated disabling these tests?  The referenced issue doesn't have any information on this.

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs [hunk 10]
```diff
@@ -106,28 +131,90 @@ public override ImmutableArray<PortableExecutableReference> ResolveReference(str
...
+            }
+
+            return null;
+        }
+
+        private static ImmutableDictionary<string, string> GetTrustedPlatformAssemblyMap()
+        {
+            var set = ImmutableDictionary.CreateBuilder<string, string>(StringComparer.OrdinalIgnoreCase);
+
+            if (CoreClrShim.AppContext.GetData?.Invoke("TRUSTED_PLATFORM_ASSEMBLIES") is string paths)
```
Comment (jaredpar) [inline src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: Can you provide some documentation on this value?  I have an idea of what it does but I want to read up more about it.  It's possible this may help me with some code in the test caching framework.

File: src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs (hunk 2 above)
Comment (nguerrera) [inline src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs]: I see. 

It's not important to change it, but you can get the caught exception from Assert.Throws and make further assertions about it:

``` C#
var exception = Assert.Throws<FileLoadException>(() => ...);
Assert.True(exception.InnerException is InteractiveAssemblyLoaderException);
```

File: src/Compilers/Shared/CoreClrShim.cs (hunk 7 above)
Comment (nguerrera) [inline src/Compilers/Shared/CoreClrShim.cs]: CoreCLR loader will allow this to bind to a later version, so as long as project.json only goes up from this, it will be fine. It makes sense to ask for the lowest version you actually need.

Comment (jaredpar) [inline src/Compilers/Shared/CoreClrShim.cs]: Sure.  But isn't the lowest version we need always what is in our project.json file?  

Or put it another way: if there is a change to the version in our project.json file, is there a reason to have this value be different?

File: src/Scripting/CSharpTest/CommandLineRunnerTests.cs (hunk 9 above)
Comment (tmat) [inline src/Scripting/CSharpTest/CommandLineRunnerTests.cs]: The tests only run on Desktop on Windows before the change.

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs (hunk 10 above)
Comment (tmat) [inline src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: The only "documentation" is here https://github.com/dotnet/coreclr/blob/c637c919adcd0c0d7fcd8818810bcb727408fa51/src/coreclr/hosts/coreconsole/coreconsole.cpp#L435

@gkhanna79 Can you provide more?

File: src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs (hunk 8 above)
Comment (tmat) [inline src/Interactive/EditorFeatures/Core/Extensibility/Interactive/InteractiveEvaluator.cs]: Added.

File: src/Compilers/Shared/CoreClrShim.cs (hunk 7 above)
Comment (tmat) [inline src/Compilers/Shared/CoreClrShim.cs]: The API we need here is from netstandard 1.6, which is in contract assembly version 4.1.0.0; netstandard 1.3 assembly version is 4.0.0.0 and doesn't have this API.
The package we reference in project.json (v4.3.0) contains both contract assemblies.

File: src/Scripting/CoreTestUtilities/RedirectedOutput.cs (hunk 6 above)
Comment (tmat) [inline src/Scripting/CoreTestUtilities/RedirectedOutput.cs]: @tannergooding I think this is better, since it doesn't require two separate things (attributes and usage of this class) to be always aligned.

File: src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs (hunk 10 above)
Comment (gkhanna79) [inline src/Scripting/Core/Hosting/Resolvers/RuntimeMetadataReferenceResolver.cs]: It basically defines a set of absolute paths to files that comprise the application and are logically part of the default load context (in which the app operates by default).

File: src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs (hunk 2 above)
Comment (tmat) [inline src/Scripting/CSharpTest.Desktop/InteractiveSessionTests.cs]: I see. OK, next time I'll know :)

**Comment by tmat** [comment]
@srivatsn for ask mode approval.

**Comment by tmat** [comment]
test windows_eta_open_prtest  please

**Comment by txchen** [comment]
Hi, when can we get a new build with this fix, please?

**Comment by tmat** [comment]
@txchen Now: https://dotnet.myget.org/gallery/roslyn

**Comment by txchen** [comment]
@tmat, thanks I figured out to tune my nuget.config to get from myget source, hope official release can be shipped soon.

**Comment by Iwuh** [comment]
Is there any way to know when this fix is in a stable release? I installed the prerelease for now but I'd like to move to stable when possible.
